sdk/middleware/sqhttp/http.go

```diff
@@ -44,28 +46,109 @@ import (
+		_ = m.ServeHTTP(w, r)
+	})
+}
+
+// MiddlewareWithError is a helper middleware to define other middlewares for
+// other frameworks thanks to the error returned by the handlers in order
+// to know if a request is being aborted.
+func MiddlewareWithError(next Handler) Handler {
+	// TODO: move this middleware function into the agent internal package (which
+	//  needs restructuring the SDK)
+	return HandlerFunc(func(w http.ResponseWriter, r *http.Request) (err error) {
+		if err := addSecurityHeaders(w); err != nil {
+			return err
+		}
 		// Create a new sqreen request wrapper.
 		req := sdk.NewHTTPRequest(r)
 		defer req.Close()
 		// Use the newly created request compliant with `sdk.FromContext()`.
 		r = req.Request()
-
 		// Check if an early security action is already required such as based on
 		// the request IP address.
 		if handler := req.SecurityResponse(); handler != nil {
 			handler.ServeHTTP(w, r)
-			return
+			return AbortRequestError{}
 		}
-
 		// Call next handler.
-		next.ServeHTTP(w, r)
-
-		// Check if a security response should be applied now after having used
-		// `Identify()` and `MatchSecurityResponse()`.
+		err = next.ServeHTTP(w, r)
+		// If the returned error is not nil nor a security response, return it now.
+		var secResponse sdk.SecurityResponseMatch
+		if err != nil && !xerrors.As(err, &secResponse) {
+			return err
+		}
+		// Otherwise check if a security response should be applied now, after
+		// having used `Identify()` and `MatchSecurityResponse()`.
 		if handler := req.UserSecurityResponse(); handler != nil {
 			handler.ServeHTTP(w, r)
-			return
+			return AbortRequestError{}
 		}
+		return nil
 	})
 }
+
+// Handler is equivalent to http.Handler but returns an error when the request
+// should no longer be handled.
+type Handler interface {
+	ServeHTTP(w http.ResponseWriter, r *http.Request) error
+}
+
+// HandlerFunc is equivalent to http.HandlerFunc but returns an error when the
+// request should no longer be handled.
+type HandlerFunc func(http.ResponseWriter, *http.Request) error
+
+// ServeHTTP calls f(w, r).
+func (f HandlerFunc) ServeHTTP(w http.ResponseWriter, r *http.Request) error {
+	return f(w, r)
+}
+
+// AbortRequestError is returned by handlers when some security response was
+// triggered and handled the response. The request handling should therefore
+// stop.
+type AbortRequestError struct {
+	Message string
+}
+
+func (AbortRequestError) Error() string {
+	return "request aborted"
+}
+
+// addSecurityHeaders is a mean to add a hook to the function closure returned
+// by MiddlewareWithError() since it is not possible to get the symbol of
+// function closures at compilation-time, so it is not possible to create a hook
+// with the address of the function closure. The solution for this precise case
+// where only a prolog is enough is therefore to simply define a function having
+// a hook and called by the closure.
+func addSecurityHeaders(w http.ResponseWriter) (err error) {
+	{
```

Comment:
Because I try to write what will be generated later by our tool and therefore defining another lexical scope in order to avoid conflicts with the one of the function.